README.md
<p align="center">  
    <img alt="" src="/images/surface_preview.gif" />
</p>

# Aerial - Apple TV Screen Saver for Windows
 
Aerial is a Windows screen saver that mimics the Apple TV screen saver that displays the videos Apple shot over New York, San Francisco, Hawaii, China, etc. as well as from the ISS and various underwater locations.

Aerial is inspired by [Mac Aerial Screen Saver](https://github.com/JohnCoates/Aerial) and [Aerial for Windows](https://github.com/cDima/Aerial).

[Complete list of available videos.](https://aerial-screensavers.netlify.com)

Aerial is open source so feel free to contribute.

## Installing
>🚧 This project is still in beta. It might not always work or behave as expected. 🚧

Download the latest release from [here](https://github.com/OrangeJedi/Aerial/releases). Follow the instructions to install Aerial.

##### [Version History](https://github.com/OrangeJedi/Aerial/wiki/Version-History)

## Features
* Cycles through randomly selected Apple TV videos
* Easily select only the video you want to play
* Control video playback settings (speed, brightness, contrast, hue, saturation, etc.)
* Skip videos with the right arrow key
* Multi-Screen Support
    * Play a different video on each screen
    * Play the same video on each screen
* Display text
  * Static text
  * Date and time
  * Display the location/description of the video
* Change the font, size, and color of text
* Sync to the time of day (mostly - not all videos have a 'day' or 'night' attribute yet)

## To-Do
* Full Multi-Screen Support
* Display text
  * Sunrise & sunset
  * Custom image
  * Weather
  * Calendar and events?
* Cache videos
* Import your own videos
* Video transitions
* 4K Videos
* Video preview in Window's screensaver settings
* Auto update

## Want to Contribute?

Check out our [guide to contributing](/CONTRIBUTING.md).

## About
This is a [Node.JS](https://nodejs.org)/[Electron](https://www.electronjs.org/) based implementation of [Mac Aerial Screen Saver](https://github.com/JohnCoates/Aerial) for Windows. I was using cDima's [Aerial for Windows](https://github.com/cDima/Aerial) on my computer but wished it had more of the features found in the MacOS version, as well as being less buggy and having better 4K support. Not knowing much about C# and with the project seemingly dead I researched Windows screensavers and found I could make a screen saver with Node.JS and Electron – both of which I am very familiar with. Hopefully I’ll be able build on this implementation and it get it to near where the MacOS version is.

## License
[MIT License](https://github.com/OrangeJedi/Aerial/blob/master/LICENSE)


documentation/changelog.md
# Changelog

## [Beta v0.4.0](https://github.com/OrangeJedi/Aerial/releases/tag/v0.4.0-beta) - March 26, 2020

* Display text on the screen saver
    * Choose from 9 placed to put text
    * Choose between text types
        * Static Text
        * Date/Time
            * Format with moment.js string
        * Video Information
            * Name
            * Accessibility label
            * Points of interest
    * Change the font, font size, and text color for all text, or just one position
* Updates to video information

## [Beta v0.3.0](https://github.com/OrangeJedi/Aerial/releases/tag/v0.3.0-beta) - March 18, 2020

* Prioritize current time of day
* Use the right arrow key to skip videos
* Play the same video on each screen
* Change the playback speed
* Video now plays on an HTML5 Canvas
* Added Video Filters
    * Blur
    * Brightness
    * Grayscale
    * Hue Rotate
    * Invert
    * Saturate
    * Sepia
* Videos scale properly on non 16:9 monitors

## [Beta v0.2.0](https://github.com/OrangeJedi/Aerial/releases/tag/v0.2.0-beta) - March 14, 2020

* Preview Videos in the Config menu
* Select which videos should play
    * Select/Deselect videos based on type

## [Beta v0.1.0](https://github.com/OrangeJedi/Aerial/releases/tag/v0.1.0-beta) - March 12, 2020

First Release!

* Cycles through a random Apple TV video
* Displays a different video on each monitor
* Toggle text in the bottom right corner with the date and time


app.js
const {app, BrowserWindow, ipcMain, screen, shell} = require('electron');
const videos = require("./videos.json");
const Store = require('electron-store');
const store = new Store();
const request = require('request');
const https = require('https');
const fs = require('fs');
let screens = [];
let nq = false;
const cachePath = `${app.getPath('userData')}/videos`;
let downloading = false;

//make video cache directory
if (!fs.existsSync(`${app.getPath('userData')}/videos/`)) {
    fs.mkdirSync(`${app.getPath('userData')}/videos/`);
}

global.shared = {
    currentlyPlaying: ''
};

function createConfigWindow(argv) {
    let win = new BrowserWindow({
        width: 1000,
        height: 750,
        webPreferences: {
            nodeIntegration: true
        }
    });
    win.loadFile('web/config.html');
    win.on('closed', function () {
        win = null;
    });
    if (argv.includes("/dt")) {
        win.webContents.openDevTools();
    }
}

function createJSONConfigWindow() {
    let win = new BrowserWindow({
        width: 1920,
        height: 1080,
        webPreferences: {
            nodeIntegration: true
        }
    });
    win.loadFile('json-editor/index.html');
    win.on('closed', function () {
        win = null;
    });
}

function createSSWindow() {
    let displays = screen.getAllDisplays();
    for (let i = 0; i < screen.getAllDisplays().length; i++) {
        let win = new BrowserWindow({
            width: displays[i].size.width,
            height: displays[i].size.height,
            webPreferences: {
                nodeIntegration: true
            },
            x: displays[i].bounds.x,
            y: displays[i].bounds.y,
            fullscreen: true,
            transparent: true,
            frame: false
        });
        win.setMenu(null);
        win.loadFile('web/screensaver.html');
        win.on('closed', function () {
            xWin = null;
        });
        win.webContents.on('dom-ready', (event) => {
            let css = '* { cursor: none !important; }';
            win.webContents.insertCSS(css);
        });
        screens.push(win);
    }
}

function createSSPWindow(argv) {
    let win = new BrowserWindow({
        width: 1920,
        height: 1080,
        webPreferences: {
            nodeIntegration: true
        },
        /*transparent: true,
        frame: false*/
    });
    win.loadFile('web/screensaver.html');
    win.on('closed', function () {
        win = null;
    });
    if (argv.includes("/dt")) {
        win.webContents.openDevTools();
    }
}

app.whenReady().then(startUp);

function startUp() {
    if (!store.get("configured")) {
        let allowedVideos = [];
        for (let i = 0; i < videos.length; i++) {
            allowedVideos.push(videos[i].id);
        }
        store.set('allowedVideos', allowedVideos);
        store.set('downloadedVideos', []);
        store.set('alwaysDownloadVideos', []);
        store.set('neverDownloadVideos', []);
        store.set('timeOfDay', false);
        store.set('sunrise', "06:00");
        store.set('sunset', "18:00");
        store.set('playbackSpeed', 1);
        store.set('skipVideosWithKey', true);
        store.set("configured", true);
        store.set('videoFilters', [
            {name: 'blur', value: 0, min: 0, max: 100, suffix: "px", defaultValue: 0},
            {name: 'brightness', value: 100, min: 0, max: 100, suffix: "%", defaultValue: 100},
            {name: 'grayscale', value: 0, min: 0, max: 100, suffix: "%", defaultValue: 0},
            {name: 'hue-rotate', value: 0, min: 0, max: 360, suffix: "deg", defaultValue: 0},
            {name: 'invert', value: 0, min: 0, max: 100, suffix: "%", defaultValue: 0},
            {name: 'saturate', value: 100, min: 0, max: 256, suffix: "%", defaultValue: 100},
            {name: 'sepia', value: 0, min: 0, max: 100, suffix: "%", defaultValue: 0},
        ]);
        store.set('sameVideoOnScreens', false);
        store.set('timeString', "dddd, MMMM Do YYYY, h:mm:ss a");
        store.set('textFont', "Segoe UI");
        store.set('textSize', "2");
        store.set('textColor', "#FFFFFF");
        store.set('displayText', {
            'positionList': ["topleft", "topright", "bottomleft", "bottomright", "left", "right", "middle", "topmiddle", "bottommiddle"],
            'topleft': {'type': "none", "defaultFont": true},
            'topright': {'type': "none", "defaultFont": true},
            'bottomleft': {'type': "none", "defaultFont": true},
            'bottomright': {'type': "none", "defaultFont": true},
            'left': {'type': "none", "defaultFont": true},
            'right': {'type': "none", "defaultFont": true},
            'middle': {'type': "none", "defaultFont": true},
            'topmiddle': {'type': "none", "defaultFont": true},
            'bottommiddle': {'type': "none", "defaultFont": true}
        });
        store.set('videoProfiles', []);
        store.set('videoTransitionLength', 2000);
        store.set('videoCache', false);
        store.set('videoCacheProfiles', false);
        store.set('videoCacheSize', getCacheSize());
        store.set('videoCacheRemoveUnallowed', false);
        store.set('cachePath', cachePath);
    }
    if (process.argv.includes("/nq")) {
        nq = true;
    }
    if (store.get('videoCacheRemoveUnallowed')) {
        removeAllUnallowedVideosInCache();
    }
    removeAllNeverAllowedVideosInCache();
    if (process.argv.includes("/c")) {
        createConfigWindow(process.argv);
    } else if (process.argv.includes("/p")) {
        //createSSPWindow();
        app.quit();
    } else if (process.argv.includes("/s")) {
        createSSWindow();
    } else if (process.argv.includes("/t")) {
        createSSPWindow(process.argv);
    } else if (process.argv.includes("/j")) {
        createJSONConfigWindow();
    } else {
        createConfigWindow();
    }
    setTimeout(downloadVideos, 1500);
}

//let Aerial load the video with the self-signed cert
app.on('certificate-error', (event, webContents, url, error, certificate, callback) => {
    if (url.match(/^https:\/\/sylvan.apple.com/) !== null) {
        event.preventDefault();
        callback(true)
    } else {
        callback(false)
    }
});

ipcMain.on('quitApp', (event, arg) => {
    if (!nq) {
        app.quit();
    }
});

ipcMain.on('keyPress', (event, key) => {
    if (key === "ArrowRight" && store.get('skipVideosWithKey')) {
        for (let i = 0; i < screens.length; i++) {
            screens[i].webContents.send('newVideo');
        }
    } else {
        if (!nq) {
            app.quit();
        }
    }
});

ipcMain.on('updateCache', (event) => {
    const path = `${app.getPath('userData')}/videos`;
    let videoList = [];
    fs.readdir(path, (err, files) => {
        files.forEach(file => {
            if (file.includes('.mov')) {
                videoList.push(file.slice(0, file.length - 4));
            }
        });
        if (!downloading) {
            store.set('downloadedVideos', videoList);
        }
        store.set('videoCacheSize', getCacheSize());
        event.reply('displaySettings');
    });
});

ipcMain.on('deleteCache', (event) => {
    removeAllVideosInCache();
    event.reply('displaySettings');
});

ipcMain.on('openCache', (event) => {
    shell.openExternal(cachePath);
});

//file download
function downloadFile(file_url, targetPath, callback) {
    // Save variable to know progress
    var received_bytes = 0;
    var total_bytes = 0;

    agentOptions = {
        host: 'sylvan.apple.com'
        , path: '/'
        , rejectUnauthorized: false
    };

    let agent = new https.Agent(agentOptions);

    let req = request({
        method: 'GET',
        uri: file_url,
        agent: agent
    });

    let out = fs.createWriteStream(targetPath);
    req.pipe(out);

    req.on('response', function (data) {
        // Change the total bytes value to get progress later.
        total_bytes = parseInt(data.headers['content-length']);
    });

    req.on('data', function (chunk) {
        // Update the received bytes
        received_bytes += chunk.length;

        showProgress(received_bytes, total_bytes);
    });

    req.on('end', function (e) {
        //console.log("File successfully downloaded");
        callback();
    });

    req.on('error', function (err) {
        //console.error(err)
    });

    function showProgress(received, total) {
        let percentage = (received * 100) / total;
        //console.log(percentage + "% | " + received + " bytes out of " + total + " bytes.");
    }
}

function downloadVideos() {
    let allowedVideos = getVideosToDownload();
    let downloadedVideos = store.get('downloadedVideos') ?? [];
    let flag = false;
    for (let i = 0; i < allowedVideos.length; i++) {
        if (!downloadedVideos.includes(allowedVideos[i])) {
            let flag = true;
            let index = videos.findIndex((v) => {
                if (allowedVideos[i] === v.id) {
                    return true;
                }
            });
            //console.log(`Downloading ${videos[index].name}`);
            downloadFile(videos[index].src.H2641080p, `${cachePath}/${allowedVideos[i]}.mov`, () => {
                downloadedVideos.push(allowedVideos[i]);
                store.set('downloadedVideos', downloadedVideos);
                store.set('videoCacheSize', getCacheSize());
                downloadVideos();
            });
            break;
        }
    }
    downloading = flag;
}

function getAllFilesInCache() {
    return fs.readdirSync(cachePath);
}

function getCacheSize() {
    let totalSize = 0;
    getAllFilesInCache().forEach(function (filePath) {
        if (fs.existsSync(`${cachePath}/${filePath}`)) {
            totalSize += fs.statSync(`${cachePath}/${filePath}`).size;
        }
    });
    return totalSize;
}

function removeAllVideosInCache() {
    getAllFilesInCache().forEach(file => {
        if (fs.existsSync(`${cachePath}/${file}`)) {
            fs.unlink(`${cachePath}/${file}`, (err) => {
            });
        }
    });
    store.set('videoCacheSize', getCacheSize());
}

function removeAllUnallowedVideosInCache() {
    let allowedVideos = getVideosToDownload();
    let downloadedVideos = store.get('downloadedVideos') ?? [];
    for (let i = 0; i < downloadedVideos.length; i++) {
        if (!allowedVideos.includes(downloadedVideos[i])) {
            fs.unlink(`${cachePath}/${downloadedVideos[i]}.mov`, (err) => {
            });
        }
    }
    updateVideoCache();
}

function removeAllNeverAllowedVideosInCache() {
    let neverAllowedVideos = store.get('neverDownloadVideos');
    let downloadedVideos = store.get('downloadedVideos') ?? [];
    for (let i = 0; i < downloadedVideos.length; i++) {
        if (neverAllowedVideos.includes(downloadedVideos[i])) {
            fs.unlink(`${cachePath}/${downloadedVideos[i]}.mov`, (err) => {
            });
            downloadedVideos.splice(i, 1);
            i--;
        }
    }
    store.set('videoCacheSize', getCacheSize());
}

function getVideosToDownload() {
    let allowedVideos = store.get('videoCache') ? store.get('allowedVideos') : [];
    store.get('alwaysDownloadVideos').forEach(e => {
        allowedVideos.push(e);
    });
    if (store.get("videoCacheProfiles") && store.get('videoCache')) {
        store.get('videoProfiles').forEach(e => {
            allowedVideos.push(...e.videos);
        });
    }
    allowedVideos = allowedVideos.filter(function (item, pos, self) {
        return self.indexOf(item) === pos;
    });
    store.get('neverDownloadVideos').forEach(e => {
        if (allowedVideos.includes(e)) {
            allowedVideos = allowedVideos.splice(allowedVideos.indexOf(e), 1);
        }
    });
    return allowedVideos;
}

function updateVideoCache() {
    let videoList = [];
    fs.readdir(cachePath, (err, files) => {
        files.forEach(file => {
            if (file.includes('.mov')) {
                videoList.push(file.slice(0, file.length - 4));
            }
        });
        if (!downloading) {
            store.set('downloadedVideos', videoList);
        }
        store.set('videoCacheSize', getCacheSize());
    });
}

json-editor/index.js
const {ipcRenderer} = require('electron');
let videos = require("../videos.json");
const beautify = require("json-beautify");

const Store = require('electron-store');
const store = new Store();

document.getElementById('output').value = beautify(videos, null, 2,128);

function beautifyInput() {
    document.getElementById('input').value = beautify(JSON.parse(document.getElementById('input').value.replace(/(\/\*([\s\S]*?)\*\/)|(\/\/(.*)$)/gm, '')), null, 2,96);
}

function createJSON() {
    let newData = JSON.parse(document.getElementById('input').value);
    if(Array.isArray(newData.assets)){
        newData = newData.assets;
    }
    if(!Array.isArray(newData)){
        alert("Error: No Array Found");
        return null;
    }
    let list = [];
    for(let i = 0;i< newData.length;i++){
        list.push({
            'id' : newData[i].id,
            'accessibilityLabel' : newData[i].accessibilityLabel,
            'src' : {
                'H2641080p' : newData[i]["url-1080-H264"]
            }
            }
        )
    }
    document.getElementById('output').value = beautify(list.sort(sortVideos),null,2,128);
    videos = list;
}

function updateJSON() {
    let newData = JSON.parse(document.getElementById('input').value);
    for (const vid in newData) {
        let index = videos.findIndex((e) => {
            if(vid === e.id){
                return true;
            }
        });
        if(index > -1) {
            videos[index].name = newData[vid].name;
            if (typeof newData[vid].pointsOfInterest === "object") {
                videos[index].pointsOfInterest = newData[vid].pointsOfInterest;
            }
        }
    }
    document.getElementById('output').value = beautify(videos.sort(sortVideos),null,2,128);
}

function addToJSON() {
    let newData = JSON.parse(document.getElementById('input').value);
    if(!Array.isArray(newData)){
        alert("Error: No Array Found");
        return null;
    }
    for (let i = 0; i < newData.length;i++) {
        let index = videos.findIndex((e) => {
            if(newData[i] === e.id){
                return true;
            }
        });
        if(index > -1) {
            videos[index][document.getElementById('attr').value] = document.getElementById('attrValue').value;
        }
    }
    document.getElementById('output').value = beautify(videos.sort(sortVideos),null,2,128);
}

function addPropertyToJSON() {
    let newData = JSON.parse(document.getElementById('input').value);
    for (const vid in newData) {
        console.log(vid);
        let index = videos.findIndex((e) => {
            if(vid === e.id){
                return true;
            }
        });
        if(index > -1) {
            videos[index][document.getElementById('attr').value] = newData[vid];
        }
    }
    document.getElementById('output').value = beautify(videos.sort(sortVideos),null,2,128);
}

function sortVideos(a,b){
    var nameA = a.accessibilityLabel.toUpperCase(); // ignore upper and lowercase
    var nameB = b.accessibilityLabel.toUpperCase(); // ignore upper and lowercase
    if (nameA < nameB) {
        return -1;
    }
    if (nameA > nameB) {
        return 1;
    }

    // names must be equal
    return 0;
}

web/screensaver.js
const {ipcRenderer, remote} = require('electron');
const videos = require("../videos.json");
const Store = require('electron-store');
const store = new Store();
const allowedVideos = store.get("allowedVideos");
let downloadedVideos = store.get("downloadedVideos");
let currentlyPlaying = '';
let poiTimeout, transitionTimeout;

function quitApp() {
    ipcRenderer.send('quitApp');
}

ipcRenderer.on('newVideo', () => {
    newVideo();
});

//quit when a key is pressed
document.addEventListener('keydown', (e) => {
    ipcRenderer.send('keyPress', e.code);
});
document.addEventListener('mousedown', quitApp);
setTimeout(function () {
    var threshold = 5;
    document.addEventListener('mousemove', function (e) {
        if (threshold * threshold < e.movementX * e.movementX
            + e.movementY * e.movementY) {
            quitApp();
        }
    });
}, 1500);

function randomInt(min, max) {
    return Math.floor(Math.random() * max) - min;
}

let video = document.getElementById("video");

video.addEventListener('play', (event) => {
    video.style.backgroundColor = "black";
});
video.addEventListener('ended', (event) => {
    newVideo();
});

function newVideo() {
    clearTimeout(poiTimeout);
    clearTimeout(transitionTimeout);
    videoAlpha = 0;
    let id = "";
    if (store.get('timeOfDay')) {
        let time = getTimeOfDay();
        id = tod[time][randomInt(0, tod[time].length)];
    } else {
        id = allowedVideos[randomInt(0, allowedVideos.length)];
    }
    if (store.get('sameVideoOnScreens')) {
        if (currentlyPlaying === remote.getGlobal('shared').currentlyPlaying) {
            remote.getGlobal('shared').currentlyPlaying = id;
        } else {
            id = remote.getGlobal('shared').currentlyPlaying;
        }
    }
    let index = videos.findIndex((e) => {
        if (id === e.id) {
            return true;
        }
    });
    let videoInfo = videos[index];
    downloadedVideos = store.get("downloadedVideos");
    let videoSRC = videoInfo.src.H2641080p;
    if(downloadedVideos.includes(videoInfo.id)){
        videoSRC = `${store.get('cachePath')}/${videoInfo.id}.mov`;
    }
    video.src = videoSRC;
    video.playbackRate = Number(store.get('playbackSpeed'));
    currentlyPlaying = videoInfo.id;
    video.onplay = onVideoPlay;
    //display text
    for (let position of displayText.positionList) {
        let textArea = $(`#textDisplay-${position}`);
        if (displayText[position].type === "information") {
            if (displayText[position].infoType === "poi") {
                changePOI(position, -1, videoInfo["pointsOfInterest"]);
            } else {
                textArea.text(videoInfo[displayText[position].infoType]);
            }
        }
    }
}

let transitionLength = store.get('videoTransitionLength');
let videoAlpha = 1;

function onVideoPlay(e) {
    fadeVideoIn(transitionLength);
    setTimeout(fadeVideoOut, (e.target.duration * 1000) - transitionLength - 300, transitionLength);
}

function fadeVideoOut(time) {
    if (time > 0) {
        transitionTimeout = setTimeout(fadeVideoOut, 5, time - 5);
    }
    videoAlpha = time / transitionLength;
}

function fadeVideoIn(time) {
    if (time > 0) {
        transitionTimeout = setTimeout(fadeVideoIn, 5, time - 5);
    }
    videoAlpha = 1 - (time / transitionLength);
}

function changePOI(position, currentPOI, poiList) {
    let poiS = Object.keys(poiList);
    for (let i = 0; i < poiS.length; i++) {
        if (Number(poiS[i]) > currentPOI) {
            $(`#textDisplay-${position}`).text(poiList[poiS[i]]);
            if (i < poiS.length) {
                poiTimeout = setTimeout(changePOI, (Number(poiS[i + 1]) - Number(poiS[i])) * 1000, position, poiS[i], poiList);
            }
            break;
        }
    }
}

//time of day code
let tod = {"day": [], "night": [], "none": []};
if (store.get('timeOfDay')) {
    for (let i = 0; i < allowedVideos.length; i++) {
        let index = videos.findIndex((e) => {
            if (allowedVideos[i] === e.id) {
                return true;
            }
        });
        switch (videos[index].timeOfDay) {
            case "day":
                tod.day.push(allowedVideos[i]);
                break;
            case "night":
                tod.night.push(allowedVideos[i]);
                break;
            default:
                tod.none.push(allowedVideos[i]);
        }
        if (tod.day.length <= 3) {
            tod.day.push(...tod.none);
        }
        if (tod.night.length <= 3) {
            tod.night.push(...tod.none);
        }
    }
}

function getTimeOfDay() {
    let cHour = new Date().getHours();
    let cMin = new Date().getMinutes();
    let sunriseHour = store.get('sunrise').substring(0, 2);
    let sunriseMinute = store.get('sunrise').substring(3, 5);
    let sunsetHour = store.get('sunrise').substring(0, 2);
    let sunsetMinute = store.get('sunrise').substring(3, 5);
    let time = "night";
    if (cHour >= sunriseHour && cMin >= sunriseMinute && cHour < sunsetHour && cMin < sunriseMinute) {
        time = "day";
    }
    return time;
}

//put the video on the canvas
let c1 = document.getElementById('canvasVideo');
let ctx1 = c1.getContext('2d');
c1.width = window.innerWidth;
c1.height = window.innerHeight;
let videoFilters = store.get('videoFilters');
let filter = "";
for (let i = 0; i < videoFilters.length; i++) {
    filter += `${videoFilters[i].name}(${videoFilters[i].value}${videoFilters[i].suffix}) `;
}
ctx1.filter = filter;

function drawVideo() {
    ctx1.clearRect(0, 0, window.innerWidth, window.innerHeight);
    ctx1.globalAlpha = 1;
    ctx1.fillStyle = "#000000";
    ctx1.fillRect(0, 0, window.innerWidth, window.innerHeight);
    ctx1.globalAlpha = videoAlpha;
    ctx1.drawImage(video, 0, 0, window.innerWidth, window.innerHeight);
    requestAnimationFrame(drawVideo);
}

drawVideo();

function runClock(position, timeString) {
    $(`#textDisplay-${position}`).text(moment().format(timeString));
    setTimeout(runClock, 1000 - new Date().getMilliseconds(), position, timeString);
}

//set up css
$('.displayText').css('font-family', `"${store.get('textFont')}"`).css('font-size', `${store.get('textSize')}vw`).css('color', `${store.get('textColor')}`);

//draw text
let displayText = store.get('displayText');
let html = "";

//create content divs
for (let position of displayText.positionList) {
    let align = "";
    if (position.includes("left")) {
        align = "w3-left-align"
    } else if (position.includes("middle")) {
        align = "w3-center"
    } else if (position.includes("right")) {
        align = "w3-right-align"
    }
    html += `<div class="w3-display-${position} ${align} w3-container textDisplayArea" id="textDisplay-${position}" style="text-shadow:.05vw .05vw 0 #444"></div>`;
    $('#textDisplayArea').html(html);
}
//add text to the content
for (let position of displayText.positionList) {
    switch (displayText[position].type) {
        case "none":
            break;
        case "text":
            $(`#textDisplay-${position}`).text(displayText[position].text);
            break;
        case "time":
            runClock(position, displayText[position].timeString);
            break;
    }
    if (!displayText[position].defaultFont) {
        $(`#textDisplay-${position}`).css('font-family', `"${displayText[position].font}"`).css('font-size', `${displayText[position].fontSize}vw`).css('color', `${displayText[position].fontColor}`);
    }
}

//play a video
newVideo();